server/src/main/java/org/tctalent/server/service/db/impl/SavedSearchServiceImpl.java

```diff
@@ -137,1562 +135,1584 @@
+   * <p/>
+   * Has to be annotated as Transactional in order to create the "persistence context" (what the
+   * underlying Hibernate calls a Session). This context is used to fetch lazily loaded attributes
+   * (by auto generating other SQL calls on the database).
+   * <p/>
+   * When running searches from requests through the REST API, Spring automatically creates this
+   * context - so you don't have to annotate all your REST API methods as Transactional. - JC
+   */
+  //Midnight GMT
+  @Scheduled(cron = "0 1 0 * * ?", zone = "GMT")
+  @SchedulerLock(name = "SavedSearchService_notifySearchWatchers", lockAtLeastFor = "PT23H", lockAtMostFor = "PT23H")
+  @Transactional
+  public void notifySearchWatchers() {
+    String currentSearch = "";
+    try {
+      Set<SavedSearch> searches = savedSearchRepository.findByWatcherIdsIsNotNull();
+      Map<Long, Set<SavedSearch>> userNotifications = new HashMap<>();
+
+      log.info("Notify watchers: running " + searches.size() + " searches");
+
+      int count = 0;
+
+      OffsetDateTime yesterday = OffsetDateTime.now().minusDays(1);
+
+      //Look through all watched searches looking for any that have candidates that were
+      //created since yesterday.
+      //Those are the searches that need to notify their watchers.
+      for (SavedSearch savedSearch : searches) {
+
+        count++;
+        currentSearch = savedSearch.getName() + " (" + savedSearch.getId() + ")";
+        log.info("Running search " + count + ": " + currentSearch);
+
+        SearchCandidateRequest searchCandidateRequest = convertToSearchCandidateRequest(
+            savedSearch);
+
+        //Set up paging
+        searchCandidateRequest.setPageNumber(0);
+        //Short page is all we need - we are only going to look at first element
+        searchCandidateRequest.setPageSize(1);
+
+        Page<Candidate> candidates = doSearchCandidates(searchCandidateRequest);
+
+        boolean newCandidates = false;
+        if (candidates.getNumberOfElements() > 0) {
+          //Get first (latest) candidate
+          Candidate candidate = candidates.getContent().getFirst();
+          OffsetDateTime createdDate = candidate.getCreatedDate();
+          newCandidates = createdDate.isAfter(yesterday);
+        }
+
+        if (newCandidates) {
+          //Query has new results. Need to let watchers know
+          Set<Long> watcherUserIds = savedSearch.getWatcherUserIds();
+          for (Long watcherUserId : watcherUserIds) {
+            Set<SavedSearch> userWatches = userNotifications.computeIfAbsent(watcherUserId,
+                k -> new HashSet<>());
+            userWatches.add(savedSearch);
           }
         }
-     */
+      }
 
-        User user = userService.getLoggedInUser();
+      //Construct and send emails
+      for (Long userId : userNotifications.keySet()) {
+        final Set<SavedSearch> savedSearches = userNotifications.get(userId);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.

Suggested remediation:
Access the values and keys with `Map.entry.getValue()` and `Map.entry.getKey()`, respectively, after iterating over the Map.entrySet().
```
@@ -1089,4 +1089,5 @@
       //Construct and send emails
-      for (Long userId : userNotifications.keySet()) {
-        final Set&lt;SavedSearch&gt; savedSearches = userNotifications.get(userId);
+      for (Map.Entry&lt;Long, Set&lt;SavedSearch&gt;&gt; entry : userNotifications.entrySet()) {
+        Long userId = entry.getKey();
+        final Set&lt;SavedSearch&gt; savedSearches = entry.getValue();
         String s = savedSearches.stream().map(SavedSearch::getName)

```